### PR TITLE
DEP: Update pybids dependency to test sqlalchemy refactor

### PR DESCRIPTION
**Pull request by effigies**
See #138.

@adelavega You can test with the `dep/pybids_sqlalchemy` branch on the `poldracklab/fitlins` repository.

### CODE REVIEW

**Comment by effigies** [comment]
With #141 merged, some superficial errors can be quickly found.

First up:

```
Traceback (most recent call last):
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1224, in _execute_context
    cursor, statement, parameters, context
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 549, in do_executemany
    cursor.executemany(statement, parameters)
sqlite3.IntegrityError: UNIQUE constraint failed: associations.src, associations.dst

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/plugins/multiproc.py", line 69, in run_node
    result['result'] = node.run(updatehash=updatehash)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 473, in run
    result = self._run_interface(execute=True)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 557, in _run_interface
    return self._run_command(execute)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 637, in _run_command
    result = self._interface.run(cwd=outdir)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 375, in run
    runtime = self._run_interface(runtime)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/fitlins/interfaces/bids.py", line 181, in _run_interface
    ignore=ignore, derivatives=derivatives)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/layout/layout.py", line 226, in __init__
    index_metadata=index_metadata)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/layout/layout.py", line 556, in add_derivatives
    self.derivatives[pipeline_name] = BIDSLayout(deriv, **kwargs)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/layout/layout.py", line 211, in __init__
    indexer.index_metadata()
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/bids/layout/index.py", line 330, in index_metadata
    self.session.commit()
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 1026, in commit
    self.transaction.commit()
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 493, in commit
    self._prepare_impl()
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 472, in _prepare_impl
    self.session.flush()
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2451, in flush
    self._flush(objects)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2589, in _flush
    transaction.rollback(_capture_exception=True)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 129, in reraise
    raise value
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2549, in _flush
    flush_context.execute()
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 422, in execute
    rec.execute(self)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 589, in execute
    uow,
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 245, in save_obj
    insert,
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 1066, in _emit_insert_statements
    c = cached_connections[connection].execute(statement, multiparams)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/sql/elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1248, in _execute_context
    e, statement, parameters, cursor, context
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1466, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 383, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 128, in reraise
    raise value.with_traceback(tb)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1224, in _execute_context
    cursor, statement, parameters, context
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 549, in do_executemany
    cursor.executemany(statement, parameters)
sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: associations.src, associations.dst
[SQL: INSERT INTO associations (src, dst, kind) VALUES (?, ?, ?)]
[parameters: (('/data/task-rhymejudgment_bold.json', '/data/sourcedata/sub-01/func/sub-01_task-rhymejudgment_bold.nii.gz', 'Metadata'), ('/data/sourcedata/sub-01/func/sub-01_task-rhymejudgment_bold.nii.gz', '/data/task-rhymejudgment_bold.json', 'Metadata'), ('/data/sourcedata/task-rhymejudgment_bold.json', '/data/task-rhymejudgment_bold.json', 'Child'), ('/data/task-rhymejudgment_bold.json', '/data/sourcedata/task-rhymejudgment_bold.json', 'Parent'), ('/data/task-rhymejudgment_bold.json', '/data/sourcedata/sub-02/func/sub-02_task-rhymejudgment_bold.nii.gz', 'Metadata'), ('/data/sourcedata/sub-02/func/sub-02_task-rhymejudgment_bold.nii.gz', '/data/task-rhymejudgment_bold.json', 'Metadata'), ('/data/sourcedata/task-rhymejudgment_bold.json', '/data/task-rhymejudgment_bold.json', 'Child'), ('/data/task-rhymejudgment_bold.json', '/data/sourcedata/task-rhymejudgment_bold.json', 'Parent')  ... displaying 10 of 104 total bound parameter sets ...  ('/data/task-rhymejudgment_bold.json', '/data/sub-13/func/sub-13_task-rhymejudgment_space-T1w_desc-preproc_bold.nii.gz', 'Metadata'), ('/data/sub-13/func/sub-13_task-rhymejudgment_space-T1w_desc-preproc_bold.nii.gz', '/data/task-rhymejudgment_bold.json', 'Metadata'))]
(Background on this error at: http://sqlalche.me/e/gkpj)
```

Looks like building a layout on https://datasets.datalad.org/?dir=/labs/poldrack/ds003_fmriprep is failing. Note that I haven't retrieved all of the images, if that makes a difference.

cc @tyarkoni

**Comment by tyarkoni** [comment]
Ah, looks like for some reason the same associations between files are being registered more than once. Should be an easy fix.

**Comment by effigies** [comment]
Okay. Note that this is a derivative dataset with the original dataset in `sourcedata/`, in case that's hitting an edge case.

**Comment by tyarkoni** [comment]
To expand, I'm guessing that for some reason multiple kinds of associations (e.g., both inheritance and metadata relationships) are being added for a given pair of files. This isn't actually illegal in BIDS—for example, a given file could be both the parent (via inheritance) of another and have an `IntendedFor` field in its sidecar, in which case it's technically associated in two different ways. I guess the solution is to impose the uniqueness constraint on all three fields—i.e., source file, destination file, and relationship kind, and not just the first two. Should have a fix momentarily.

**Comment by tyarkoni** [comment]
Okay, I pushed a commit that should (fingers crossed) fix this.

File: requirements.txt [hunk 1]
```diff
@@ -4,5 +4,5 @@ numpy>=1.11
 nilearn>=0.4.0
 pandas>=0.19
 nipype>=1.1.6
-pybids>=0.8.0
+git+https://github.com/bids-standard/pybids.git@2bcdd85c5d6370e9a3cfe79ddcba5fec7434704d#egg=pybids
```
Comment (effigies) [inline requirements.txt]:
```suggestion
git+https://github.com/bids-standard/pybids.git@2af465080656c9869297f8fb44275497a2fb5a98#egg=pybids
```

**Comment by effigies** [comment]
@tyarkoni I invited you as a collaborator. If you're trying to iterate on pybids, feel free to suggest a new githash on the review page and apply it, as I did in https://github.com/poldracklab/fitlins/pull/139#discussion_r280128534. That will probably be faster than waiting on me.

**Comment by effigies** [comment]
Same problem.

**Comment by tyarkoni** [comment]
Okay, will look into it more carefully when I can. You can try initializing with `reset_database=True`, though I doubt that will help unless you were already setting `database_file` to something (as it should be using an in-memory DB by default, so there's nothing to reset).

**Comment by tyarkoni** [comment]
Ah, nevermind—I see the source of the problem. Not sure *why* it's happening, but will look into it.

File: requirements.txt [hunk 2]
```diff
@@ -4,5 +4,5 @@ numpy>=1.11
 nilearn>=0.4.0
 pandas>=0.19
 nipype>=1.1.6
-pybids>=0.8.0
+git+https://github.com/bids-standard/pybids.git@2af465080656c9869297f8fb44275497a2fb5a98#egg=pybids
```
Comment (tyarkoni) [inline requirements.txt]:
```suggestion
git+https://github.com/bids-standard/pybids.git@cc462c009c33fa44bdbb3b8661a5b937e079802e#egg=pybids
```

File: requirements.txt [hunk 3]
```diff
@@ -4,5 +4,5 @@ numpy>=1.11
 nilearn>=0.4.0
 pandas>=0.19
 nipype>=1.1.6
-pybids>=0.8.0
+git+https://github.com/bids-standard/pybids.git@cc462c009c33fa44bdbb3b8661a5b937e079802e#egg=pybids
```
Comment (tyarkoni) [inline requirements.txt]:
```suggestion
git+https://github.com/bids-standard/pybids.git@0ac80d0647420fda677da3cca9180ecb24bcb878#egg=pybids
```

**Comment by tyarkoni** [comment]
Okay, that problem is solved...

**Comment by effigies** [comment]
Looks like some syntax changes are needed on my end. I'll see to those. Thanks for the pybids help.

**Comment by tyarkoni** [comment]
Hooray!

**Comment by effigies** [comment]
Looks like this damages building the reports, which depends on a `BIDSLayout` to find the relevant files.

**Comment by tyarkoni** [comment]
You could try replacing all `extensions` mentions with `extension`, and dropping any leading periods. In theory these are deprecated and not broken, and both should be handled internally for now, but it's possible something isn't working quite right (and you may as well future-proof now anyway).

**Comment by effigies** [comment]
Does `extension` accept a single string, or does it need to be a list of strings? I'm currently passing a single, and the problem might be that `extensions` now expects a list instead of automatically promoting to a list.

**Comment by tyarkoni** [comment]
Unless I'm mistaken, that shouldn't matter... all entities can handle lists implicity. All I'm doing to handle `extensions` is popping it off and renaming it. And the leading period is also handled with listification. The relevant code:

```python
        if 'extensions' in filters:
            filters['extension'] = filters.pop('extensions')
            warnings.warn("In pybids 0.9.0, the 'extensions' filter was "
                          "deprecated in favor of 'extension'. The former will"
                          " stop working in 0.11.0.")

        # For consistency with past versions where "extensions" was a
        # hard-coded argument, allow leading periods
        if 'extension' in filters:
            exts = listify(filters['extension'])
            filters['extension'] = [x.lstrip('.') for x in exts]
```

**Comment by tyarkoni** [comment]
Have not looked closely, but latest error(s) are probably because the `extension` is already extracted from the filename somewhere else (it's now a normal entity) and is then being added a second time.

**Pull request closed without merging**